pkg/agent/multicluster/mc_route_controller.go

```diff
@@ -211,39 +238,192 @@ func (c *MCRouteController) Run(stopCh <-chan struct{}) {
+	defer c.ciImpQueue.Done(obj)
+
+	ciImp, ok := obj.(string)
+	if !ok {
+		c.ciImpQueue.Forget(obj)
+		klog.InfoS("Expected string in work queue but got", "object", obj)
+		return true
+	}
+	if c.wireGuardConfig.Enable {
+		if err := c.updatePeers(ciImp); err != nil {
+			c.ciImpQueue.AddRateLimited(ciImp)
+			klog.ErrorS(err, "Error updating WireGuard peers, requeuing", "key", ciImp)
+		}
+	}
+
+	if err := c.syncMCFlows(); err == nil {
+		c.ciImpQueue.Forget(obj)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.ciImpQueue.AddRateLimited(ciImp)
+		klog.ErrorS(err, "Error syncing key, requeuing", "key", ciImp)
+	}
+	return true
 }
 
-func (c *MCRouteController) processNextWorkItem() bool {
-	obj, quit := c.queue.Get()
+func (c *MCRouteController) processNextGatewayItem() bool {
+	obj, quit := c.gatewayQueue.Get()
 	if quit {
 		return false
 	}
-	defer c.queue.Done(obj)
+	defer c.gatewayQueue.Done(obj)
 
-	if k, ok := obj.(string); !ok {
-		c.queue.Forget(obj)
+	gateway, ok := obj.(string)
+	if !ok {
+		c.gatewayQueue.Forget(obj)
 		klog.InfoS("Expected string in work queue but got", "object", obj)
 		return true
-	} else if err := c.syncMCFlows(); err == nil {
-		c.queue.Forget(k)
+	}
+
+	if c.wireGuardConfig.Enable && gateway == c.nodeConfig.Name {
+		if err := c.reconcileWireGuard(gateway); err != nil {
+			klog.ErrorS(err, "Error when reconcile WireGuard, requeuing", "key", gateway)
+			c.gatewayQueue.AddRateLimited(gateway)
+		}
+	}
+	if err := c.syncMCFlows(); err == nil {
+		c.gatewayQueue.Forget(obj)
 	} else {
 		// Put the item back on the workqueue to handle any transient errors.
-		c.queue.AddRateLimited(k)
-		klog.ErrorS(err, "Error syncing key, requeuing", "key", k)
+		c.gatewayQueue.AddRateLimited(gateway)
+		klog.ErrorS(err, "Error syncing key, requeuing", "key", gateway)
 	}
 	return true
 }
 
+func (c *MCRouteController) reconcileWireGuard(gateway string) error {
+	if !c.wireGuardConfig.Initialized {
+		klog.InfoS("Initializing WireGuard interface and rule")
+		if err := c.initializeWireGuardInterface(); err != nil {
+			return err
+		}
+		if err := c.initHostRoute(); err != nil {
+			return err
+		}
+		c.wireGuardConfig.Initialized = true
+		return nil
+	}
+
+	if _, err := c.mcClient.MulticlusterV1alpha1().Gateways(c.namespace).Get(context.TODO(), gateway, metav1.GetOptions{}); err != nil {
+		if !apierrors.IsNotFound(err) {
+			return fmt.Errorf("failed to get Gateway %s, error: %s", gateway, err.Error())
+		}
+		klog.InfoS("Gateway is deleted, removing WireGuard interface and rule")
+		if err := c.deleteHostRoute(); err != nil {
+			return err
+		}
+		c.wireGuardConfig.Initialized = false
+	}
+
+	return nil
+}
+
+func (c *MCRouteController) deleteHostRoute() error {
+	if err := c.routeClient.DeleteMulticlusterWireGuardRule(int(openflow.MulticlusterPktMark), multiclusterWireGuardTable); err != nil {
+		return err
+	}
+	link, err := c.routeClient.LinkByName(multiclusterWireGuardInterface)
```

Comment:
I don't if it is graceful to delete a route by deleting its corresponding interface. I would delete the route only. Another question: since the name of the method is `deleteHostRoute`, why deleting the interface?

Reply:
Sure, maybe we can rename this method to `RemoveHostStaticRouteAndIface` or others? But when this method is called, which means the Node is not Gateway anymore, so we need to clean both iface and static route.

Reply:
Agreed. Probably name it `deleteWireGuardInterface`?